BehaviorTreeEditor.cs

```diff
@@ -75,16 +188,97 @@ void CreateTypeLists()
 				delegate( Type type ) { return type.ToString(); } ) );
 	}
 
-	public static Type CreateNodeTypeSelector( TreeNode node )
+	public static void DeleteNode( TreeNode parentNode )
 	{
-		int selectedType = ( node != null ? Array.IndexOf<Type>( nodeTypes, node.GetType() ) : 0 );
-		selectedType = EditorGUILayout.Popup( "<---->", selectedType, nodeTypeNames );
+		if ( parentNode is Decorator )
+		{
+			DeleteNode( ( (Decorator)parentNode )._child );
+		}
+		else if ( parentNode is Compositor )
+		{
+			foreach ( TreeNode child in ( (Compositor)parentNode )._children )
+			{
+				DeleteNode( child );
+			}
+		}
+
+		DestroyImmediate( parentNode, true );
+	}
+
+	public Type CreateNodeTypeSelector( TreeNode node, int height )
+	{
+		int selectedType = ( node != null ? Array.IndexOf<Type>( nodeTypes, node.GetType() ) : Array.IndexOf<Type>( nodeTypes, typeof( NullNode ) ) );
+		Rect rect = new Rect( _indentLevel * _indentSize + _foldoutWidth, height * _itemHeight, 200, _itemHeight );
+		selectedType = EditorGUI.Popup( rect, selectedType, nodeTypeNames );
 		return nodeTypes[selectedType];
 	}
 
+	public bool CreateChildrenFoldout( TreeNode node, int height )
+	{
+		if ( !_editorData.ContainsKey( node ) )
+		{
+			_editorData.Add( node, new EditorNode() );
+		}
+
+		EditorNode nodeData = _editorData[node];
+
+		Rect rect = new Rect( _indentLevel * _indentSize, ( height - 1 ) * _itemHeight, _foldoutWidth, _itemHeight );
+		nodeData.foldout = EditorGUI.Foldout( rect, nodeData.foldout, "children", true );
+		return nodeData.foldout;
+	}
+
 	public static TreeNode CreateNode( Type nodeType )
 	{
-		return (TreeNode)Activator.CreateInstance( nodeType );
+		DebugUtils.Assert( _instance, "Cannot create a node without a current editor instance!" );
+		DebugUtils.Assert( _instance._behaviorTree, "Can't create a node without an asset to add it to!" );
+
+		TreeNode newNode = (TreeNode)ScriptableObject.CreateInstance( nodeType );
+		AssetDatabase.AddObjectToAsset( newNode, _instance._behaviorTree );
+		EditorUtility.SetDirty( _instance._behaviorTree );
+		return newNode;
 	}
+
+	public static TreeNode nullNode
+	{
+		get
+		{
+			return CreateNode( typeof( NullNode ) );
+		}
+	}
+
+	public static BehaviorTree CloneTree( BehaviorTree behaviorTree )
+	{
+		BehaviorTree treeClone = Instantiate<BehaviorTree>( behaviorTree );
+		treeClone.root = CloneNode( behaviorTree.root );
+		return treeClone;
+	}
+
+	public static TreeNode CloneNode( TreeNode node )
+	{
+		TreeNode nodeClone = Instantiate<TreeNode>( node );
+
+		if ( node is Decorator )
+		{
+			( (Decorator)nodeClone )._child = CloneNode( ( (Decorator)node )._child );
+		}
+		else if ( node is Compositor )
+		{
+			List<TreeNode> cloneChildren = new List<TreeNode>();
+			foreach ( TreeNode child in ( (Compositor)node )._children )
+			{
+				cloneChildren.Add( (TreeNode)CloneNode( child ) );
+			}
+
+			( (Compositor)nodeClone )._children = cloneChildren;
+		}
```

Comment:
you do this if decorator, else if compositor stuff a couple places. consider adding a virtual method that returns a list of children instead. decorator returns a list of length one. might be less efficient but could make these segments easier to read.